### PR TITLE
Add parameter server train & side-car eval on k8s

### PR DESCRIPTION
**Pull request by selcukgun**
ResNet56 model (with custom training loop) variables are created on
parameter server jobs, and updated by workers. Evaluation is done using
a dedicated job which uses the checkpoints saved during the training
(side-car evaluation).

The model is trained on CIFAR10 dataset.

### CODE REVIEW

File: distribution_strategy/parameter_server_training/README.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
+# Parameter Server Training Using Distribution Strategies
+
+This directory provides an example of running parameter server training with
+Distribution Strategies.
+
+Please first read the
+[documentation](https://www.tensorflow.org/tutorials/distribute/parameter_server_training)
+of Distribution Strategy for parameter server training. We also assume that readers
+of this page  are familiar with [Google Cloud](https://cloud.google.com/) and
```
Comment (yuefengz) [inline distribution_strategy/parameter_server_training/README.md:9]: Redundant space

File: distribution_strategy/parameter_server_training/README.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+its [Kubernetes Engine](https://cloud.google.com/kubernetes-engine/).
+
+This directory contains the following files:
+
+-   kubernetes/template.yaml.jinja: jinja template used for generating Kubernetes manifests
+-   kubernetes/render_template.py: script for rendering the jinja template
+-   Dockerfile.resnet_cifar_ps_strategy: a docker file to build the model image
+-   resnet_cifar_ps_strategy.py: script for running any type of parameter server training task based on `TF_CONFIG` environment variable
```
Comment (yuefengz) [inline distribution_strategy/parameter_server_training/README.md:17]: "any type of ..." seems too general, maybe just say "a ResNet example using Cifar dataset for parameter server training"

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    "Whether to use in-process cluster to run training or evaluation.")
+
+FLAGS = flags.FLAGS
+
+TRAIN_EPOCHS = 182
+STEPS_PER_EPOCH = 781
+BATCH_SIZE = 64
+EVAL_BATCH_SIZE = 8
+
+def create_in_process_cluster(num_workers, num_ps):
```
Comment (yuefengz) [inline distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:53]: Could you update the work_config part according to this tutorial? https://www.tensorflow.org/tutorials/distribute/parameter_server_training#in-process_cluster

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+def train_resnet_cifar(cluster_resolver):
+  """Trains the resnet56 model using parameter server distribution strategy.
+
+  Args:
+      cluster_resolver: cluster resolver to give neccessary information to
+        set up distributed training
+  """
+
+  strategy = parameter_server_strategy_v2.ParameterServerStrategyV2(
```
Comment (yuefengz) [inline distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:104]: let's use tf.distribute.experimental.ParameterServerStrategy.

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+                     TRAIN_EPOCHS):
+    per_worker_train_iterator = iter(per_worker_train_dataset)
+    per_worker_eval_iterator = iter(per_worker_eval_dataset)
+    for _ in range(STEPS_PER_EPOCH):
+      coordinator.schedule(worker_train_fn, args=(per_worker_train_iterator,))
+    coordinator.join()
+    logging.info("Finished joining at epoch %d. Training accuracy: %f.",
+                  epoch, train_accuracy.result())
+
+    for _ in range(STEPS_PER_EPOCH):
```
Comment (yuefengz) [inline distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:221]: Should evaluation use a different steps_per_epoch? since you have a different batch_size for evaluation.

Comment (yuefengz) [inline distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:221]: Could you add a comment here saying that we are running inline distributed evaluation, in this case an evaluator job is not necessary.

**Review by yuefengz — COMMENTED** [review]
Thank for your PR!

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py (hunk 5 above)
Comment (selcukgun) [reply distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:221]: Good point. Introducing EVAL_STEPS_PER_EPOCH and setting it to 88 in the next patch shortly. This gives us a probability of 0.99 for a row in the dataset to be evaluated.

File: distribution_strategy/parameter_server_training/README.md (hunk 1 above)
Comment (selcukgun) [reply distribution_strategy/parameter_server_training/README.md]: Done.

File: distribution_strategy/parameter_server_training/README.md (hunk 2 above)
Comment (selcukgun) [reply distribution_strategy/parameter_server_training/README.md]: Done.

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py (hunk 3 above)
Comment (selcukgun) [reply distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:54]: Added inter ops.

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py (hunk 4 above)
Comment (selcukgun) [reply distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py]: Done.

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py (hunk 5 above)
Comment (selcukgun) [reply distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py]: Done.

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+from tensorflow_models.official.benchmark.models import cifar_preprocessing
+from tensorflow_models.official.benchmark.models import resnet_cifar_model
+from tensorflow_models.official.vision.image_classification.resnet import common as img_class_common
+
+flags.DEFINE_string("checkpoint_dir", "gs://cifar10_ckpt/",
+                    "Directory for writing model checkpoints.")
+flags.DEFINE_string("data_dir", "gs://cifar10_data/",
+                    "Directory for Resnet Cifar model input. Follow the "
+                    "instruction here to get Cifar10 data: "
+                    "https://github.com/tensorflow/models/tree/r1.13.0/official/resnet#cifar-10")
```
Comment (yuefengz) [inline distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:36]: redundant new line?

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+          ).assert_existing_objects_matched()
+
+    train_dataset_fn = lambda _: cifar_preprocessing.input_fn(
+        is_training=True,
+        data_dir=FLAGS.data_dir,
+        batch_size=BATCH_SIZE,
+        parse_record_fn=cifar_preprocessing.parse_record,
+        dtype=tf.float32,
+        drop_remainder=True)
+    eval_dataset_fn = lambda _: cifar_preprocessing.input_fn(
```
Comment (yuefengz) [inline distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:190]: Is the eval data shuffled? If not, could you add a comment and a TODO?

**Review by yuefengz — APPROVED** [review]

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py (hunk 7 above)
Comment (yuefengz) [reply distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:190]: Maybe you can just append a shuffle at the end of the dataset?

Comment (selcukgun) [reply distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:190]: input_fn already shuffles the training data using process_record_dataset: [code link](https://github.com/tensorflow/models/blob/871c4e0a393ef4385534bee55354a5df8aa1ccf4/official/legacy/image_classification/resnet/imagenet_preprocessing.py#L108)

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py (hunk 6 above)
Comment (selcukgun) [reply distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:36]: Split the help argument into multiple lines for readability; they are displayed as concatenated if help cmdline arg is passed.

File: distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    per_worker_eval_iterator = iter(per_worker_eval_dataset)
+    for _ in range(STEPS_PER_EPOCH):
+      coordinator.schedule(worker_train_fn, args=(per_worker_train_iterator,))
+    coordinator.join()
+    logging.info("Finished joining at epoch %d. Training accuracy: %f.",
+                  epoch, train_accuracy.result())
+
+    # Since we are running inline evaluation below, a side-car evaluator job is not necessary.
+    for _ in range(EVAL_STEPS_PER_EPOCH):
+      coordinator.schedule(worker_eval_fn, args=(per_worker_eval_iterator,))
```
Comment (rainwoodman) [inline distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py:230]: We can probably build a similar API for DTensor async training. A major difficulty to sort out is what to do if worker_eval_fn( and or replica_fn) is multi-mesh -- for example if there is a summary Op that needs to run on a the CPU.
